[Doc][AMDGPU] Add barrier execution & memory model

Pierre-vh opened the pull request:
Add a formal execution model, and a memory model for the execution barrier
primitives available in GFX12.0 and below.
The model also works for GFX12.5 workgroup/workgroup trap barriers, but does
not include the new barrier types and instructions added in GFX12.5.
These will be added at a later date.

Pierre-vh committed:
[Doc][AMDGPU] Add barrier execution & memory model

Add a formal execution model, and a memory model for the execution barrier
primitives available in GFX12.0 and below.
The model also works for GFX12.5 workgroup/workgroup trap barriers, but does
not include the new barrier types and instructions added in GFX12.5.
These will be added at a later date.

nhaehnle commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6553,6 +6567,281 @@ The Private Segment Buffer is always requested, but the Private Segment
+  orders:
+
+  * *Thread-barrier-order<BO>* for every barrier object ``BO``.
+  * *Barrier-participates-in<BO>* for every barrier object ``BO``.
```
Partial orders are transitive by definition.

The union of these two total orders is not transitive.

So, something has to give there. Do we want barrier-executes-before to be the transitive closure or not?

nhaehnle reviewed (commented):
Side note: It feels like it may be time to split up AMDGPUUsage.rst. It's become a bit unwieldy.

Overall, I really like this.

Pierre-vh commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6553,6 +6567,281 @@ The Private Segment Buffer is always requested, but the Private Segment
+  barrier *object* ``BO``, then all of the following is true:
+
+  * ``S`` cannot be empty. :sup:`WIP`
+  * The elements of ``S`` are all ordered by a continuous interval of *barrier-modification-order<BO>*.
```
Yes exactly. Maybe "ordered by" is too much, and I can just say "are in" ?

Pierre-vh commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6553,6 +6567,281 @@ The Private Segment Buffer is always requested, but the Private Segment
+
+When a barrier *signal* ``S`` *barrier-executes-before* a barrier *wait* ``W``, ``S`` executes before ``W``
+**as-if** ``S`` is *program-ordered* before ``W``. Thus, all *dynamic instances* *program-ordered* before ``S``
+are known to have been executed before the *dynamic instances* *program-ordered* after ``W``.
```
Indeed, this is one of the paragraphs I struggled the most with. Your understanding is correct, it *must not* imply happens-before.

Perhaps I should move this "as-if" to the informational notes section ?

nhaehnle replied:
Yes, that would be better. Or just remove it entirely? I'm not sure what it's supposed to say.

Pierre-vh commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6678,6 +6967,30 @@ following sections:
+* ``S`` *barrier-executes-before* ``W``.
+* *BA* and *BR*'s :ref:`synchronization scope<amdgpu-memory-scopes>` overlap.
+* *BA* and *BR*'s :ref:`synchronization scope<amdgpu-memory-scopes>`
+  allow cross address space synchronization (they cannot have ``one-as``).
```
> What is this last rule for?

The theoretical reason is that barrier objects are stored in an unspecified memory region which may or may not overlap with any known address space, so we demand cross-address-space synchronization to allow pairing with that special barrier AS.

The practical reason is that the waits we insert for `one-as` are softer because we can take advantage of the properties of each AS in that mode. For example we don't need to wait on LDS for a workgroup-one-as release fence, but for a fence to pair with a barrier, we really need the "cross-address-space" synchronization mode where we always require the ops to complete so they can be observed from another AS.

> (I've always felt that the one-as scopes are a bit weird and we should probably supersede them with something else, so this is more out of curiosity.)

Agreed

Pierre-vh committed:
Comments

Pierre-vh reviewed (commented):
> Side note: It feels like it may be time to split up AMDGPUUsage.rst. It's become a bit unwieldy.

Agreed, I thought about doing it here but I feel like it's better done as a separate task. There needs to be a discussion on what "granularity" we want the split to be done.

nhaehnle reviewed (commented):
Thanks. Most of my remaining comments are more of an editorial nature. The one comment I would like to see addressed is the one about linking waves and threads.

Pierre-vh committed:
Comments

nhaehnle commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6553,6 +6567,292 @@ The Private Segment Buffer is always requested, but the Private Segment
+*Barrier-executes-before* represents the order in which barrier operations will complete by relating the
+dynamic instances of operations from different threads together.
+For example, if ``A -> B`` in *barrier-executes-before*, then the execution of ``A`` must complete
+before the execution of ``B`` can complete..
```
Double period at the end.

nhaehnle commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6553,6 +6567,297 @@ The Private Segment Buffer is always requested, but the Private Segment
+    The *expected count* of the barrier object is set to the number of waves in the workgroup
+  * The number of waves in the workgroup is not known until all waves of the workgroup have launched.
+    Thus, the *expected count* of *workgrou barrier object* can never be equal to its *signal count*
+    until all wavefronts of the workgroup launched.
```
> > barrier operations can only appear in wave-uniform control flow, and
> 
> I believe the statement about `Barrier operations are convergent within a wave` is already enough to guarantee this. @ssahasra is that correct or do I need another statement to explicitly forbid using a barrier within non-uniform control flow ?

At least I didn't read the statement in this way. The word "convergent" is really only used as an adjective describing a static property of an operation.

(I notice one point of ConvergenceAndUniformity.rst there is a mention of "... threads that execute convergent operations convergently", but that should probably be changed to "... treads that execute convergent operations converge**dly**" or something like that.)

IMHO the clearest way is just to say "Barrier operations can only appear in wave-uniform control flow."

A downside of that wording is that we don't actually seem to define uniform control flow in ConvergenceAndUniformity.rst, but that could be easily fixed by saying something along the lines of "a thread executes a dynamic instance of an operation is in X-uniform control flow if it is converged with dynamic instances from all other threads of X".

Pierre-vh replied:
I updated the wording on that, and I also added the following bullet point:

```
  * Thus, barrier operations can only be performed in wave-uniform control flow.
```

I believe that is a good starting point, we can improve it further over time

ssahasra replied:
I need to page in the rest of the review, and I am okay with this wording in the short term. But when we get around to actually defining "wave-uniform control flow", we should move away from anything that creates the impression of "threads are executing lock-step" ... that is the naive picture that a real programmer has in their mind, and it can often lead to wrong conclusions such as "but I made sure that the threads converged N steps ago". "Executing together" or "dynamic instances are converged" should strictly be limited to dynamic instances and not "flow of dynamic instances".

Pierre-vh committed:
Comments

Pierre-vh committed:
Fix rules about barrier init, and prevent out of thin air behavior

Pierre-vh commented:
I updated the specification to fix the following:

- Rules around barrier init were imprecise.
- There were not enough safeguards to prevent cyclic execution ("out of thin air" barrier completion). I added a requirement for the modification order to be consistent with executes-before, and also added an informational note about it.

Pierre-vh committed:
Clarify barrier-modification-order

nhaehnle reviewed (approved):
Editing nits inline, but I think this is pretty much ready to go.

Pierre-vh committed:
Comments

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+
+Threads can synchronize execution by performing barrier operations on barrier *objects* as described below:
+
+* Barrier *objects* have the following state:
```
```suggestion
* Each barrier *object* has the following state:
```

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+
+* Barrier *objects* have the following state:
+
+  * An unsigned non-zero positive integer *expected count*: counts the number of *signal* operations
```
```suggestion
  * *expected count*: An unsigned non-zero positive integer. Counts the number of *signal* operations
```
And similarly for signal count. Perhaps make this a definition list?

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+    * There is no barrier *leave* ``L`` where ``J -> L -> X`` in *thread-barrier-order<BO>*.
+
+  * *Barrier-participates-in<BO>* is a partial order that relates barrier operations to barrier *waits*.
+    A barrier operation ``X`` may *barrier-participates-in<BO>* a barrier *wait* ``W`` if all of the following
```
```suggestion
    A barrier operation ``X`` *barrier-participates-in<BO>* a barrier *wait* ``W`` if all of the following
```

Pierre-vh replied:
That's intentional, for a `barrier-participates-in<BO>` rel to be valid, all of this must be met, but there are also additional constraints below (with the "Let `S` ..." paragraph).

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+
+  * *Barrier-participates-in<BO>* is consistent with *happens-before*.
+
+* Let ``S`` be the set of barrier operations that *barrier-participates-in<BO>* a barrier *wait* ``W`` for some
```
```suggestion
* Let ``S`` be the set of barrier operations that *barrier-participate-in<BO>* a barrier *wait* ``W`` for some
```
I think it should be okay to use English grammar rules on the order name, especially when that name is already being used as a verb (it's being "verbed"?).

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+  * The elements of ``S`` all exist in a continuous interval of *barrier-modification-order<BO>*.
+  * Let ``A`` be the first operation of ``S`` in *barrier-modification-order<BO>*, then the *signal count* of ``BO``
+    is zero before ``A`` is performed.
+  * Let ``B`` be the last operation of ``S`` in *barrier-modification-order<BO>*, then the *signal count* and
```
Is it illegal (or UB) for a wave to indefinitely wait on a barrier because there are insufficient signal/leave operations?

Separately, this assumes that the model is used to describe an execution where the barrier is known to have completed. What if the model is being used to describe the state of the execution frozen at a point "before" the wait completes? Maybe we could change this to say that "B0 is completed at W if and only if there there is an operation in S that makes the counts equal".

Pierre-vh replied:
> Is it illegal (or UB) for a wave to indefinitely wait on a barrier because there are insufficient signal/leave operations?

I don't think so, your program just hangs like an infinite loop (more or less).

> What if the model is being used to describe the state of the execution frozen at a point "before" the wait completes?

This is not something that I ever considered while writing this model. I do not think it'd add value to the model to support this use case.

We could do it later, but I'd like to first focus on getting this across the finish line as a baseline instead of adding new things to it. We will refine it over time.

> Maybe we could change this to say that "B0 is completed at W if and only if there there is an operation in S that makes the counts equal".

I really don't think it's that easy. If you add that, then it means you allow modeling instances of executions where a wait is the last thing that happens in a thread because it hangs. That affects everything else (e.g. your thread can now end with a wait, without ever leaving the barrier; then we need to support that without allowing weird cases in "normal" executions, etc.)

ritter-x2a replied:
> I don't think so, your program just hangs like an infinite loop (more or less).

Isn't that UB? At least it is in C++, LLVM IR doesn't seem to have a very definitive answer to that.

I think we need to make barrier waits that never stop waiting UB so that the bullet point discussed here makes sense. Otherwise, executions that end mid-wait would be valid, so the behavioral model would need to allow them (which I don't think it should).

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+
+* For every barrier *init* ``I`` performed on a barrier *object* ``B0``:
+
+  * There is no barrier operation ``X`` such that ``X -> I`` in *thread-barrier-order<BO>*. :sup:`WIP`
```
I think this is a target-specific detail. More interesting is to say that for every `X`, there is an `I` preceding it in thread-barrier-order, i.e., it is illegal to operate on an uninitialized barrier.

Pierre-vh replied:
> More interesting is to say that for every X, there is an I preceding it in thread-barrier-order, i.e., it is illegal to operate on an uninitialized barrier.

This does not work because the init could be in a different thread, what we could say however is that there has to be an init that happens-before any other barrier operation on the barrier.

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+
+* For every barrier *leave* ``L`` performed on a barrier *object* ``BO``:
+
+  * There is no other barrier operation *thread-barrier-ordered<BO>* after ``L``. :sup:`WIP`
```
This is also target-specific. The more important constraint is that a wait requires a join before it. Also, a thread can leave a barrier so that it is removed from expected count, but the same thread can then wait on the barrier, right?

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+
+  * There is no other barrier operation *thread-barrier-ordered<BO>* after ``L``. :sup:`WIP`
+
+* *Barrier-executes-before* is a strict partial order of all barrier operations. It is the transitive closure of all
```
```suggestion
* *Barrier-executes-before* is a strict partial order defined over the  union of all barrier operations performed by all threads on all barriers. It is the transitive closure of all
```

ssahasra commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+
+*Barrier-executes-before* represents the order in which barrier operations will complete by relating the
+dynamic instances of operations from different threads together.
+For example, if ``A -> B`` in *barrier-executes-before*, then the execution of ``A`` must complete
```
Seems redundant. It's already an order, and this sentence seems to create an intuitive interpretation of time on that order. At best, it can be an informational note.

Pierre-vh commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,309 @@ The Private Segment Buffer is always requested, but the Private Segment
+    A barrier *join* ``J`` is *barrier-joined-before<BO>* a barrier operation ``X`` if and only if all
+    of the following is true:
+
+    * ``J -> X`` in *thread-barrier-order<BO>*.
```
Yes

Pierre-vh committed:
Address comments

Pierre-vh committed:
Fix init rule

Pierre-vh committed:
Fix RST Build Error

ritter-x2a commented:
> > I think if we'd use the term "happens-before" in LLVM, we would need to stick to the relation defined here: https://llvm.org/docs/LangRef.html#memory-model-for-concurrent-operations
> That would mean that if an initialized shared variable x is written in one thread A, then A barrier-synchronizes with another thread B, and B reads x after the barrier-sync (and no other accesses happen), B must see the value written by A (which is not the intention, without additional fences/synchronization, as I understood it).
> 
> Right. I wonder if I should add `happens-before` to `barrier-executes-before` and wherever I'd use `happens-before`, I instead use `barrier-executes-before` ?

It took a bit of getting used to for me, but I think that makes sense. The point of `barrier-executes-before` would then be to capture all cases where the programmer can be sure that one thing has executed before the other, which can be obtained through barrier synchronization, through program order (in hb), or through memory synchronization (via release/acquire, in hb).
And this kind of executes-before relation should probably also be sufficient to separate barrier inits from subsequent barrier ops.

Pierre-vh commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6548,6 +6562,305 @@ The Private Segment Buffer is always requested, but the Private Segment
+    *expected count* of ``BO`` are equal after ``B`` is performed. ``B`` is the only barrier operation in ``S``
+    that causes the *signal count* and *expected count* of ``BO`` to be equal.
+
+* For every barrier operation ``X``, there is a barrier *init* ``I`` such that ``I`` *happens-before* ``X``.
```
I moved it lower because it rewrote it to use `barrier-executes-before`, and I'm hesitant to do a forward-reference to it. What do you think?

Pierre-vh reviewed (commented):
Note: I have to use the graphite UI because GitHub gave up on the review.

Pierre-vh committed:
Comments

Pierre-vh committed:
Add informational node about waits

Pierre-vh committed:
Final offline review

Pierre-vh reviewed (commented):
We did a final review offline during a meeting, and the remaining reviewers agreed that this can land.
